### PR TITLE
Add search functionality to Client

### LINKED ISSUE #33: Implement Globus Search Utilities in Client
We need to add some utilities to simplify using Globus Search from the DLHubClient.

See https://github.com/materials-data-facility/forge and https://globus-sdk-python.readthedocs.io/en/stable/clients/search/ for inspiration.

### LINKED ISSUE #44: Add tests for Search helpers
#41 adds helpers to search for models, using Globus Search by way of MDF Forge. These helpers should have tests made for them, included with the existing dlhub_sdk tests.

Not all of the fields required for the helpers to work (such as `datacite.creators.creatorName` and `dlhub.owner`) are implemented in any existing entries, so until that is finished passing tests cannot be made. Additionally, I'm not familiar enough with the models and data in DLHub to easily create appropriate tests.

### PR DESCRIPTION
**Pull request by jgaff**
Adds search functions to client

Fixes #44 and fixes #33

### CODE REVIEW

**Comment by WardLT** [comment]
A list of things to add in for an initial release:

- [ ] Given model name, return all information for the latest version of the model
- [ ] Given model name, return the available versions
- [ ] Given model name, return the functions (names, inputs, outputs)
- [ ] Given user name, return the latest versions of all models
- [ ] Given doi, return associated models

These are all the ones I can think up for now, and will add in more later as they occur to me. I would not add any kind of rendering / reformatting of the output of these functions. We can delegate routines for printing to stdout to the CLI and HTML-friendliness to the portal.

**Comment by jgaff** [comment]
It seems like there are only a couple of search fields here: `model_name`, `user_name`, and `doi`. Those can definitely be searched for (as long as the information is in the Search entry). I'm not sure I would recommend filtering fields from the search results in the helper itself, however. If you're intending to do additional formatting later in the process, the fields of interest could be pulled out of the results at that stage. Alternatively, if you think users will want only the information you listed here (specifically the functions of a model, for example, and nothing else about the model), we could add helpers that filter the entries after they are fetched - something like `dlhc.match_model_name("my_model").search().get_versions()` or similar.

Do either of those solutions seem reasonable to you?

**Comment by WardLT** [comment]
I agree, generally, about avoiding doing filtering or any other kinds of data transformations in the search helper functions, but would consider getting only the function information an exception. I'd envision that a common task when using a servable would be to get the input and output formats, and it would great if we could do that without presenting the user with unneeded information (DataCite isn't concise). 

Also, were you envisioning the factory-method like pattern of (`dlhc.match.search()`) for these helper functions? I think it would be nice to make the helper functions just return the results eagerly, but am curious to know your take on that.

**Comment by jgaff** [comment]
If getting just the function information will be a very common case, it could be a flag in the arguments (so `dlhc.match_model_name("my_model").search(functions_only=True)` or similar would return the function information of the latest version of the model and no other information). I don't think that's a big issue, if there's only one case where this filtering would take place.

The paradigm of Forge (which I'm applying here) is that the user builds a query using `match` helpers, and then executes the query with `search`. This pattern (`dlhc.match_X().match_Y().match_Z().search()`) allows the user to add as many or as few terms to the query as desired, without needing to understand the underlying query string syntax. On top of the `match` functions, there are some premade queries (like `dlhc.search_by_authors()`) for convenience and/or example.

I'm happy to create `search_by` methods for each `match` method, but I think the `match` methods are important to expose. If a user wants to search by two fields ("I want the models from Dr. Smith in materials"), they cannot just use a premade search - they must use two `match` methods and then `search` (unless we create `search_by` methods that include every permutation of fields, which adds up to a whole lot of methods or arguments to methods, and neither are really good options). For the single-field-search case, a `search_by` is sufficient, but for more complex queries the `match` methods are important.

**Comment by WardLT** [comment]
Hmm, maybe we can draw the line for whether to employ the "match" approach or a "search by" approach when we are dealing with a single model or >1 model. 

For the cases of "get me information about this model," I like the idea of having a "search by" method for simplicity.

For things that involve hunting for models (e.g., get me Prof. X's models), we could employ the match approach for maximal flexibility. 

Sound reasonable? If so, we could reduce the number of helper functions needed and utilize the documentation to point to common types of queries (e.g., get me the models associated with this paper).

**Comment by jgaff** [comment]
We can have both a `search_by` and a `match` method for each field of interest; the overhead of creating multiple helper methods in this way is not significant. We could then direct users toward `search_by_model_name` instead of `match_model_name` if we expect the standard use case for `model_name` searching to be "get the information for this specific model (or models)." I'm not against having `search_by` methods, I'm just for also backing those up with `match` methods.

Also, I should ask, is `dlhub.name` 1) the "model name" field we're talking about and 2) always guaranteed to be unique (either representing one model with multiple versions or one version of one model)? If `dlhub.name` is the unique model identifier, but one `dlhbu.name` may have multiple versions, the `search_by_model_name` will need to optionally also search by version.

**Comment by jgaff** [comment]
Sure, that makes sense. It does seem like `search_by_model_name` (or whatever unique model identifier) is a definitely useful `search_by`, so I can implement that and hold off on the others.

Looking at the DLHub schema, I do see a `dlhub.id`, but it is allowed to be `null`, which wouldn't be unique. I just need to know what combination of fields identifies exactly one model, assuming we want users to be able to say "give me all the information on this (one) model."

**Comment by WardLT** [comment]
The "dlhub.id" is for internal use only (or legacy, not sure if we even use it internally anymore). The naming scheme we want to use publically is "user_name/model_name" (think GitHub), but I'm not sure how that naming convention is reflected in the search index.

**Comment by jgaff** [comment]
Noted. I think I'm just confusing myself more by digging through the DLHub schema, so I think I'll wait until we can figure out how models should be identified specifically.

**Comment by WardLT** [comment]
Fixes #33

**Comment by jgaff** [comment]
I believe the helpers in this branch now address all the search use cases that have been identified so far. (@WardLT do you agree?)

I'm not sure, but I think the Travis test failures may be due to the supplied credentials not being valid for Search, which isn't something I should/can fix.

File: dlhub_sdk/tests/test_dlhub_client.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 class TestClient(TestCase):
 
     def setUp(self):
-        self.dl = DLHubClient.login(http_timeout=10)
```
Comment (WardLT) [inline dlhub_sdk/tests/test_dlhub_client.py]: I figured out the problem with the tests. It has to do with this change here. Note that in `login` we do not specify a Client ID and in `__init__` we use the DLHub client ID. 

I used `login` when generating the credential files for the tests, not `__init__`. So, I'm thinking I need to get rid of the `login` function so we don't have two different Auth codes to maintain.

**Comment by WardLT** [comment]
Alright, I've fixed the Auth issue by resetting the credentials.

Some comments regarding toolbox: I was prompted to give credentials twice, is that needed or expected? Also, have you considered using the callback-server based flow from globus_cli in toolbox? 

Regarding #44 : Let's wait to merge this until it is tested. 

A general comment: Now that you have more of a feel for the DLHub client, how do you feel about hte possibility of making the Client a subclass of Forge? I'd prefer to avoid the extra baggage of all of the helper functions in Forge that aren't needed here, but the amount of duplicated docstrings we have in Forge makes me a little nervous. What's your take?

FYI: I believe the failing tests have to do with the DLHub server.

**Comment by jgaff** [comment]
There's a known bug in Toolbox that sometimes a user will be double-prompted. I'm not sure yet what causes this behavior, but I do know about it. The callback server flow isn't planned for use in Toolbox. There's work being done by the PS team on a more universal Globus login package (that would supplant most of the Toolbox login function), which does use that flow, so it still may come to Toolbox at a later date.

I agree, tests are needed.

I do still think it would be reasonable to make DLHC a subclass of Forge, but I also think this approach will work fine. I don't have a particularly strong opinion either way at this point. The benefits of subclassing would be that Forge methods wouldn't need duplicated docstrings or duplicated tests, and would all be available immediately. However, I think the current implementation, which essentially whitelists methods from Forge, is easy enough to continue and would reduce the confusion of a new user that tried to autocomplete a DLHC function in IPython (for example). The overhead of pulling in additional Forge methods is low, and all the methods in DLHC can be useful to end-users.

Are you concerned about the duplicated docstrings for maintenance reasons? I can certainly understand that, and the current approach would make breaking changes in Forge more dangerous, but Forge is pretty stable at this point (from a breaking-changes perspective; more helpers will almost certainly be added). Or do you have additional concerns about the current approach?

**Comment by WardLT** [comment]
Sounds good about the toolbox login approach.

I'm mostly concerned about the maintenance burden of the testing overhead. Maybe if we have breaking changes in Forge it'll be worthwhile to refactor Forge to make certain blocks of utility functions modular. We might also be able to play games with functools to do the "delegate to Forge" type operations. That said, I'm happy enough with this implementation.

**Comment by jgaff** [comment]
I understand that concern. It's an unnecessary burden to test Forge methods that are already validated by the Forge test suite. However, I'm not sure about the wisdom of having another layer between the user and Search, but I think a discussion about that would be useful (perhaps the base functions could be in Query, for example). I'd avoid too much cleverness with functools also for maintenance reasons, though, since we could have other, easier options.
Although for now, as long as you're fine with this implementation as a prototype version, we can defer the other discussions.

**Comment by WardLT** [comment]
Yep, let's keep this version and merge when we have tests.

Do you have entries in the index with enough metadata to make tests?

**Comment by jgaff** [comment]
I don't really know what's in the index at this point. Last I checked, the new/updated fields weren't in any entries, and some other fields (like authors, as `datacite.creator.creatorName`) weren't populated either.
Also, I'm not sure that I'm the best person to make tests against the index, because I'm not familiar with what data would be expected to be returned for a given query (and if/how that might change over time). I've manually tested the methods that can return data, but just to see that something looking correct is returned.

**Comment by WardLT** [comment]
Cool. I'll build the tests then, which will also have the benefit of someone else using this code before we ship it

**Comment by WardLT** [comment]
Alright, I think I've finished basic search functionality for the DLHubClient. 

To do list before merging:
- [ ] Write documentation
- [ ] Increment version number
- [ ] Add new MDF toolbox to requirements (once released)

**Comment by WardLT** [comment]
Fixes #44 
Fixes #33

**Pull request merged**